src/PostRector/Collector/UseNodesToAddCollector.php

```diff
@@ -30,32 +29,23 @@ final class UseNodesToAddCollector
     private array $useImportTypesInFilePath = [];
 
     public function __construct(
-        private readonly CurrentFileProvider $currentFileProvider,
         private readonly UseImportsResolver $useImportsResolver,
     ) {
     }
 
-    public function addUseImport(FullyQualifiedObjectType $fullyQualifiedObjectType): void
+    public function addUseImport(File $file, FullyQualifiedObjectType $fullyQualifiedObjectType): void
     {
-        /** @var File $file */
-        $file = $this->currentFileProvider->getFile();
-
-        $this->useImportTypesInFilePath[$file->getFilePath()][] = $fullyQualifiedObjectType;
+        $filePath = $file->getFilePath();
+        $this->useImportTypesInFilePath[$filePath][] = $fullyQualifiedObjectType;
```

Comment:
Actually, per-`$filePath` collection is actually not really needed, when many files apply import, memory will increase more, instead, reset the following properties,

- useImportTypesInFilePath
- constantUseImportTypesInFilePath
- functionUseImportTypesInFilePath

```php
public function reset(): void
{
      $this->useImportTypesInFilePath = [];
      $this->constantUseImportTypesInFilePath = [];
      $this->functionUseImportTypesInFilePath = [];
}
```

On append, can use direct object append:

```php
$this->useImportTypesInFilePath[] = $fullyQualifiedObjectType;
```

then, on PostRector's `UseAddingPostRector` `afterTraverse()`, eg:

```php
    public function afterTraverse(array $nodes): array
    {
          $this->UseNodesToAddCollector->reset();
           return $nodes;
     }
```

so next file will have empty for the first place.

Reply:
Do you have some measurements on these? I'd love to try it locally and see results.

I'm thinking having a single service to handle this, instead of juggling around a value object.

Reply:
I was thinking about it, but undoing due to BC break, basically, current implementation still have more memory if it append index file path, since service is shared, and if multiple files apply import, it will more and more append with index.

I will create separate alternative PR, with beware, as bc break.

Reply:
here the PR 

- https://github.com/rectorphp/rector-src/pull/6053